experiment/cherrypicker/server.go

```diff
@@ -0,0 +1,197 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"regexp"
+	"time"
+
+	uuid "github.com/satori/go.uuid"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/git"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/hook"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const pluginName = "cherrypick"
+
+var cherryPickRe = regexp.MustCompile(`(?m)^/cherrypick\s+(.+)$`)
+
+// Server implements http.Handler. It validates incoming GitHub webhooks and
+// then dispatches them to the appropriate plugins.
+type Server struct {
+	BotCredentials string
+	BotName        string
+	GitClient      *git.Client
+	GithubClient   *github.Client
+	HMACSecret     []byte
+	Logger         *logrus.Entry
+}
+
+// ServeHTTP validates an incoming webhook and puts it into the event channel.
+func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	eventType, _, payload, ok := hook.ValidateWebhook(w, r, s.HMACSecret)
+	if !ok {
+		return
+	}
+	fmt.Fprint(w, "Event received. Have a nice day.")
+
+	// Handle only issue comments
+	if eventType != "issue_comment" {
+		return
+	}
+
+	var ic github.IssueCommentEvent
+	if err := json.Unmarshal(payload, &ic); err != nil {
+		s.Logger.WithError(err).Info("Unmarshaling of Github payload failed.")
+		return
+	}
+
+	go func() {
+		if err := s.handle(ic); err != nil {
+			s.Logger.WithError(err).Info("Cherry-pick failed.")
+		}
+	}()
+}
+
+func (s *Server) handle(ic github.IssueCommentEvent) error {
+	// Only consider new comments in closed PRs.
+	if !ic.Issue.IsPullRequest() || ic.Issue.State != "closed" || ic.Action != github.IssueCommentActionCreated {
```

Comment:
Potentially useful UX here is to also handle PR merge events so that the command can be used to mean "cherry-pick this when it merges"

Reply:
done